BIP452: P2P UTXO Set Sharing (#2137)
bitcoin/bips: bip-XXXX.md

```
  BIP: ?
  Layer: Peer Services
  Title: P2P UTXO Set Sharing
  Authors: Fabian Jahr <fjahr@protonmail.com>
  Status: Draft
  Type: Specification
  Assigned: ?
  Discussion: 2026-05-06: https://groups.google.com/g/bitcoindev/c/rThmyI8ZN3Q

>   Version: 0.2.0
>   License: BSD-2-Clause

Reviewer: The Headers have a fixed order, please update to:

```suggestion
  Assigned: ?
  License: BSD-2-Clause
  Discussion: 2026-05-06: https://groups.google.com/g/bitcoindev/c/rThmyI8ZN3Q
  Version: 0.2.0
```

Author: Done

```

## Abstract

This BIP defines a P2P protocol extension for sharing full UTXO sets between peers. It introduces
a new service bit `NODE_UTXO_SET`, four new P2P messages (`getutxotree`, `utxotree`, `getutxoset`,
`utxoset`), and a chunk-hash list anchored to a Merkle root known to the requesting node, enabling

> per-chunk verification. This allows nodes to bootstrap from a recent height by obtaining the

Reviewer: I assume that the nodes are bootstrapping from scratch _to a_ recent height rather than from a recent height? Maybe:

```suggestion
per-chunk verification. This allows bootstrapping nodes to leapfrog to a recent height by obtaining the
```

Author: Ok, to me these sound like they mean the same but I am no native speaker and this is probably clearer.

required UTXO set directly from the P2P network via mechanisms such as assumeutxo.

## Motivation

The assumeutxo feature (implemented in Bitcoin Core) allows nodes to begin operating from a serialized
UTXO set while validating
historical blocks in the background. However, there is currently no canonical source for obtaining this
data. Users must either generate one themselves from a fully synced node (using `dumptxoutset` in 
Bitcoin Core), or download one from a third party.

By enabling UTXO set sharing over the P2P network, new nodes can obtain the data directly from
peers, removing the dependency on external infrastructure.

## Specification

The key words "MUST", "MUST NOT", "SHOULD", "SHOULD NOT", and "MAY" in this document are to be
interpreted as described in RFC 2119.

### Service Bit

| Name | Bit | Description |
|------|-----|-------------|

> | `NODE_UTXO_SET` | 12 (0x1000) | The node can serve complete UTXO set data for at least one height. |

Reviewer: This is in conflict with the Utreexo proposal which allocates

- Bit 12 to `NODE_UTREEXO`
- Bit 13 to `NODE_UTREEXO_ARCHIVE`

per the [BIP183](https://github.com/bitcoin/bips/pull/1923/changes#diff-c31ced08d65611df53a3920d0cd54e12aeac463e93df3e10ae31a4d0734fd6c5R481-R482) draft.

Author: Moved to take Bit 14, thanks for giving me the heads up. But I will most likely change this to use BIP 434 instead unless there are any late issues arising with it.

A node MUST NOT set this bit unless it has at least one full UTXO set available to serve.
A node signaling `NODE_UTXO_SET` MUST be capable of responding to `getutxotree` and `getutxoset`

> requests for every UTXO set it is willing to serve, including the full chunk-hash list and every

Reviewer: ```suggestion
requests for every UTXO set that it is willing to serve, including the full chunk-hash list and every
```

Author: Done

| `network_magic` | `bytes` | 4 | Network message start bytes. |
| `base_height` | `uint32_t` | 4 | Block height of the UTXO set. |
| `base_blockhash` | `uint256` | 32 | Block hash of the UTXO set. |
| `coins_count` | `uint64_t` | 8 | Total number of coins (UTXOs) in the set. |

**Body (coin data):**

Coins are grouped by transaction hash. For each group:

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `txid` | `uint256` | 32 | Transaction hash. |
| `num_coins` | `compact_size` | 1–9 | Number of outputs for this txid. |

For each coin in the group:

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `vout_index` | `compact_size` | 1–9 | Output index. |
| `coin` | `Coin` | variable | Serialized coin (varint-encoded code for height/coinbase, then compressed txout). |

Coins are ordered lexicographically by outpoint (txid, then vout index), matching the LevelDB iteration
order of the coins database.

#### Chunk Merkle Tree

The serialized UTXO set (header + body) is split into chunks of exactly 3,900,000 bytes (3.9 MB). The
last chunk contains the remaining bytes and may be smaller.

> The leaf hash for each chunk is `SHA256d(chunk_data)`. The tree is built as a balanced binary tree. When

Reviewer 2: nit: when reading this, I was wondering if missed a section. A very brief (non-title) introduction of the tree we're building before talking about its leafs would read more naturally.

Author: Improved the wording here, I hope it's clearer now.

> the number of nodes at a level is odd, the last node is duplicated before hashing the next level.

Reviewer: I guess this is safe because no `left_child` and `right_child` should ever have the same data, but I dimly remember that this construction had some downsides in the Merkle tree for transactions in a block. (Some node implementation accepted a block with the same transaction repeated, or smth?) Would it perhaps be better to have a dedicated value to add into the hash instead of repeating the `left_child`, e.g., `SHA256d(left_child || SHA256d(odd_number_of_leaves_no_right_child))`?

Author: Hm, while the situation is a bit different than with txs, it's a good point to raise and one way or the other, there should be an additional check added to the BIP. Generally, I think any such attack should be able to get caught based on the `utxotree` response. The attack you describe would make it necessary for the attack to have the same hash in the list twice. So we should be able to catch it with a duplication check of the chunk hash list in the `utxotree` response. The alternative you are suggesting should probably be to use `SHA256d("")` instead of the duplication but I still think we can not avoid doing a similar check. If an attacker sends us the list with `SHA256d("")` included at the end we need to identify it early, otherwise there could be a confusion about what the correct number of chunks is. The attacker could not send the final ghost chunk to us and we would keep asking peers for it and get disconnected most likely, failing to complete the snapshot even though we have all the data already (the same would be possible for the current approach BTW). At least using an empty hash as the sentinel value would ensure that we don't append any nonsense data to the snapshot if we get a result to the ghost chunk.

I did some (re) reading on the latest preferred approaches for merkle trees and it appears that domain separation through tagged hashing is valid (as in Taproot) or just promoting the odd leaf up a level. Of the two I much prefer the latter due to ease of implementation. I have thus changed the draft to use this.

I am still undecided if this is the best approach since I see a little bit of confusion potential as people might just assume the merkle tree works the same as the one for txs. The uniqueness check seems like a small price to pay. But the new version is cleaner. Happy to reconsider this change if anyone feels strongly about this or if I missed anything worth considering here.

Reviewer 3: The utxotree message include `data_length` which (along with the fixed 3.9MB data sizes) makes the merkle tree structure deterministic. If you make the hardcoded hash identifying the utxo set be the hash of all the data in the utxotree message, that should suffice to prevent messing around with the chunk structure?

Also, `num_chunks` is redundant given `data_length` and the 3.9MB constant. If you want to support chunks breaking at sizes other than 3.9MB, it would be better to have `3,900,000` be included in the message rather than `num_chunks` IMO.

BTW, I don't think you need a tree structure here btw; you could just hash the utxotree message (ie concatenate the chunk hashes) instead.

Author: > The utxotree message include data_length which (along with the fixed 3.9MB data sizes) makes the merkle tree structure deterministic. If you make the hardcoded hash identifying the utxo set be the hash of all the data in the utxotree message, that should suffice to prevent messing around with the chunk structure?

Yes, it does. It is another variant of a (very small) check in addition to the validation of the tree. So is checking the uniqueness of all the hashes in the list. My point was that there are several of these integrity checks that are possible to fix the malleability of the final data and even catch it early, and I don't really have much of a preference between them. But I liked the idea of changing the tree construction because that seems to solve the issue without any additional integrity check.

> Also, num_chunks is redundant given data_length and the 3.9MB constant. If you want to support chunks breaking at sizes other than 3.9MB, it would be better to have 3,900,000 be included in the message rather than num_chunks IMO.

Yeah, I actually thought about removing `num_chunks` from `utxotree` because of the redundancy, and I am not sure why I didn't do it yet, so I am doing it with the next push. My intention with this wasn't to support different chunk sizes, so I think the removal addresses your comment about the 3.9 MB number. I thought about parameterization, but I don't see much value in it, given the rationale I have in the BIP about the 3.9 MB choice. But if people find it compelling, I would be open to parameterizing this.

> BTW, I don't think you need a tree structure here btw; you could just hash the utxotree message (ie concatenate the chunk hashes) instead.

You are correct, of course! The tree structure is a relict of the inclusion proofs that were part of the `utxoset` message. I am still not fully convinced we should change to concatenated hashes, though. Merkle trees are well understood in Bitcoin and are not that much more complex than simple concatenation. Keeping Merkle trees still leaves the option open for inclusion proofs if they should be useful somehow down the line. But I don't feel strongly about this and can define a different way of combining chunk hashes if reviewers prefer that.

Reviewer 3: If you want to reuse the hash for inclusion proofs, then probably it would be best to divide chunks into individual utxos, and definitely to avoid having individual utxos split across chunks (meaning you'd need 8MB to prove some utxo's inclusion).

Author: Sorry, I meant inclusion of the chunks like I had been using in the original version of the BIP. But I thought about this a bit, if this could be useful and how much overhead this would create. I don't think it's worth it without an actual use-case though. We would give up a lot of simplicity without the uniform size of the chunks for example.

Interior nodes are computed as `SHA256d(left_child || right_child)`.

The leaves are delivered to the node in a single `utxotree` response. A node that knows
the Merkle root for a given UTXO set checks a received list of leaves by recomputing the root and
comparing. The Merkle root is the sole trust input required to verify the integrity of the received UTXO set.

`SHA256d` denotes double-SHA256: `SHA256d(x) = SHA256(SHA256(x))`.

### Messages

#### `getutxotree`

Sent to request the chunk-hash list for a specific UTXO set.

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `block_hash` | `uint256` | 32 | Block hash identifying the requested UTXO set. |

> A node that has advertised `NODE_UTXO_SET` and can serve the requested UTXO set MUST respond with

Reviewer 2: nit
```suggestion
A node that has advertised `NODE_UTXO_SET` and can serve the requested UTXO set SHOULD respond with
```

Author: I looked at many other BIPs, particularly 152 and 157 in terms of the language they use for similar situations and that's how I ended up with `MUST` here. Do you have a specific rationale for this change? I guess you are not feeling too strongly about it (nit), so I am leaving it as is for now.

`utxotree`. If the serving node cannot fulfill the request, it MUST NOT respond. The requesting
node SHOULD apply a reasonable timeout and try another peer.

#### `utxotree`

Sent in response to `getutxotree`, delivering the full chunk-hash list along with per-snapshot
metadata.

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `block_hash` | `uint256` | 32 | Block hash this data corresponds to. |
| `version` | `uint16_t` | 2 | Format version of the serialized UTXO set. |
| `data_length` | `uint64_t` | 8 | Total size of the serialized UTXO set in bytes (header + body). |
| `num_chunks` | `compact_size` | 1–9 | Number of chunks the serialized UTXO set is split into. |
| `chunk_hashes` | `uint256[]` | 32 × `num_chunks` | The ordered list of chunk hashes. |

> Upon receiving a `utxotree` message, the node MUST recompute the Merkle root from

Reviewer: Stumbled over this the first time, because “node” could apply to either side.

```suggestion
Upon receiving a `utxotree` message, the requesting node MUST recompute the Merkle root from
```

Author: Done

`chunk_hashes` and compare it against the Merkle root it knows for the corresponding UTXO set. If
the roots do not match, the node MUST discard the response and MUST disconnect the peer.

> #### `getutxoset`

Reviewer: `getutxoset` feels a bit odd, when the message actually requests a chunk. Also, Cluster Mempool makes extensive use of the term "chunk", and I was wondering whether this overlap could cause confusion in the future.

Author: Chunk is just a very generic term in computer science, [it's part of the http spec as well](https://en.wikipedia.org/wiki/Chunked_transfer_encoding) and thus we have the term appearing several times in the libevent replacement code as well (both client and server) which can not be avoided. Initially, I also wasn't the happiest about it but I just couldn't find a different term for it that felt right, especially considering how http uses the term and how that analogy seems to match pretty well. This also explains the message naming: The message transports the UTXO set, chunks is just an aspect of the transport mechanism in the http analogy. At least that's how felt most comfortable with reasoning about it. I think Cluster Mempool would have had a more wide pick of fitting terminology to chose from but that ship has sailed. Happy to still consider a renaming if anyone has a good suggestion but all the alternatives I could think of didn't seem to fit well enough. I also obviously prefer shorter naming in order to not make squeezing it in 12 characters too awkward.

Reviewer: I don’t feel that strongly about it, your response resolves it for me.

Sent to request a single chunk of UTXO set data. The requesting node MUST have received a `utxotree`

> for the corresponding UTXO set before sending this message.

Reviewer: Should there perhaps also be an addition that the serving node must not reply if it has not previously sent a `utxotree` message to the requesting peer?

Reviewer: I also checked that the Utreexo BIPs don’t use the term "utxotree", but I worry that "utreexo" and "utxotree" are extremely similar and that will cause headaches in the future.

Author: > Should there perhaps also be an addition that the serving node must not reply if it has not previously sent a utxotree message to the requesting peer?

I think this would require that the serving node keeps track of every request which opens new DoS possibilities. So I would rather not require this.

> I also checked that the Utreexo BIPs don’t use the term "utxotree", but I worry that "utreexo" and "utxotree" are extremely similar and that will cause headaches in the future.

Just like with the other naming response, overall happy to consider renaming if there is a good suggestion. I also didn't like this option too much due to the closeness to utreexo but I also couldn't find a better name. Using chunks in the msg name here was ruled out by me for the same reason I laid out above: chunks are a generic term within transport and don't have anything to do with the content of what is transported IMO. In the end both proposals deal with trees of UTXOs to some degree so its hard to differentiate. I guess if there was a unique brand for this proposal that could be used but I am not creative enough, best I got are "minishare" and "UTorrentXO" but nobody wants that :p

serialized UTXO set format to confirm it is well-formed.

### Protocol Flow

1. The requesting node identifies peers advertising `NODE_UTXO_SET`.
2. The requesting node sends `getutxotree` for the desired block hash to one or more of these peers.
3. Each peer responds with `utxotree`. The requesting node verifies the response by recomputing
   the Merkle root against a value it knows for the given UTXO set, either from a trusted source
   or by selecting a root with agreement among multiple peers.
4. The requesting node downloads chunks via `getutxoset`/`utxoset` exchanges, verifying each chunk
   against its entry in the accepted `utxotree` on receipt. On verification failure the peer is
   disconnected and download continues from another peer without losing already-verified chunks.
5. After all chunks are received, the node parses the reassembled UTXO set against the serialized
   UTXO set format to confirm that it is well-formed.

Serving nodes are free to limit the number of concurrent and repeated transfers per peer at their own
discretion to manage resource consumption.

## Rationale

**Usage of service bit 12:** Service bits allow selective peer discovery through

> DNS seeds and addr relay. Bit 12 is chosen as the next unassigned bit after `NODE_P2P_V2` (bit 11, BIP 324).

Reviewer: As noted above, your service flag should perhaps be moved to bit 14, unless the Utreexo project is willing to move instead.

Author: As mentioned above, moved to take Bit 14 but I will most likely change this to use BIP 434 instead in the next push.

**Direct request model:** Peers signal availability of UTXO sets via the `NODE_UTXO_SET`
service bit; the requesting node identifies the desired UTXO set by block hash when sending
`getutxotree`. The serving node responds only if it can serve that specific UTXO set.

**Per-chunk verification:** The chunk-hash list returned in `utxotree` enables each chunk to be verified
by direct lookup against the accepted list as it arrives, allowing immediate detection of corrupt data,
peer switching without data loss, and parallel download from multiple peers. The list itself is small

> (~80 KB for a ~10 GB set). The specified serialization is deterministic, so all honest nodes produce

Reviewer: If a node is expected to source chunks from multiple different peers, is it really necessary to receive the entire tree description of 80 KB from each of the peers?

Reviewer 3: The spec says "The requesting node sends `getutxotree` for the desired block hash to one or more of these peers." so I think the answer is "no, it is not necessary to receive the entire tree description of 80 KB from each of the peers" -- you only send requests to the number of peers you want to receive response from. Any attempt to give different responses will (should) result in them not hashing back to your known merkle root, so all valid descriptions will be identical, AIUI.

Reviewer: I read line 133,134

 > Sent to request a single chunk of UTXO set data. The requesting node MUST have received a `utxotree`
> for the corresponding UTXO set before sending this message.

as the serving node not being permitted to respond to `getutxoset` calls for a specific tree unless it previously sent a `utxotree` message to the same peer, but maybe I misinterpreted that. It seems to me that both aspects of the question should be clarified:

- Must a peer send `getutxotree` before being eligible to responses to `getutxoset` for the same tree?
- Is it necessary to retrieve the `utxotree` from multiple peers before requesting chunks?

Reviewer 3: Perhaps it would be better to send `getutxotree` to one node (repeating until you get a valid response), and then send `getutxoset` to any nodes that support utxo set sharing, with the response `utxoset <hash> <n> <empty>` indicating "i don't have that utxoset data" ? So instead of `getutxotree` / `utxotree` to establish whether a peer has the data you want, you send `getutxoset` / `utxoset` and either get an explicit nope or data you actually want?

Author: @murchandamus I tightened the language you were not happy with, primarily in the Protocol Flow section. Please let me know if it works better for you now.

@ajtowns It's an interesting suggestion. I am not convinced so far since I don't really like overloading original message semantics (its a new protocol, we may at least do it right in the beginning). I also think the explicit nope doesn't gain us that much and at the same time moves us backwards with regards to [@luke-jr 's concerns](https://github.com/bitcoin/bips/pull/2137#discussion_r3076101080). Did you have specific thoughts on this? If we would do your suggestion then I would almost feel better about adding the -info discovery approach again. But I haven't thought about it that much, I will continue to ponder this.

Reviewer: Reading the diff, I think there is still one thing that could be clearer: from the perspective of the _serving_ node, is it required that a peer has first sent `getutxotree` before it responds to `getutxoset` messages or will it respond to `getutxoset` even if the peer did not send `getutxotree`? Presumably it’s the latter, since `getutxotree` is a heavy message and you say that it can be requested only a single time, but I think the protocol is only specific on the _requester_ side so far.

```
  BIP: ?
  Layer: Peer Services
  Title: P2P UTXO Set Sharing
  Authors: Fabian Jahr <fjahr@protonmail.com>
  Status: Draft
  Type: Specification
  Assigned: ?

>   Discussion: ?

Reviewer 4: Please link here to the top post of the mail list thread, if you decide to open this PR later after discussion there.

Author: I wanted to address the comments I already got before posting and I was a bit busy for a while, will make the post now with this push.

Reviewer 4: Updated the PR description with links to the ML post and to the proposed implementation (edit: still needs updating in your header).

height. In Bitcoin Core, for example, the `hash_serialized` field is in the assumeutxo
parameters. It is computed by iterating over every coin in the set in lexicographic outpoint order and
feeding a serialized representation of each coin into a SHA256d hasher. The per-coin serialization is:

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `outpoint` | `COutPoint` | 36 | Transaction hash (32 bytes) + output index (4 bytes, little-endian). |
| `code` | `uint32_t` | 4 | `(height << 1) \| coinbase_flag`, little-endian. `height` is the block height at which the coin was created. `coinbase_flag` is 1 if the coin originates from a coinbase transaction, 0 otherwise. |
| `txout` | `CTxOut` | variable | The transaction output: amount as `int64_t` (8 bytes, little-endian) followed by the scriptPubKey serialized with its `compact_size` length prefix. |

All coin serializations are fed sequentially into a single SHA256d hasher. The resulting 32-byte digest
is the serialized hash.

### Messages

#### `getutxosetinfo`

Sent to discover which UTXO sets a peer can serve. This message has an empty payload.

> A node that has advertised `NODE_UTXO_SET` MUST respond with `utxosetinfo`. A node that has not

Reviewer 5: nit: I think it would be clearer to repeat here that we should disconnect a node that doesn't respond to our `getutxoset`, as stated above: 

> A node signaling `NODE_UTXO_SET` MUST respond to `getutxosetinfo` messages and MUST be capable of serving all UTXO sets it advertises in its `utxosetinfo` response. A node that fails to meet these
obligations SHOULD be disconnected.

Author: With the latest changes I felt it would be more appropriate to remove this general statement instead:

> A node that fails to meet these obligations SHOULD be disconnected.

Without discovery it is just harder for us to decide if the peer is misbehaving or not. It is more precise anyway to clarify this explicitly for each situation.

advertised the service bit SHOULD ignore this message.

#### `utxosetinfo`

Sent in response to `getutxosetinfo`. Lists available UTXO sets.

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `count` | `compact_size` | 1–9 | Number of available UTXO sets. |

For each available UTXO set:

| Field | Type | Size | Description |

> |-------|------|------|-------------|

Reviewer 6: Since the format has a version number, it would make sense to include it here.

Author: Done

| `height` | `uint32_t` | 4 | Block height. |
| `block_hash` | `uint256` | 32 | Block hash at that height. |
| `serialized_hash` | `uint256` | 32 | The UTXO set serialized hash. |
| `data_length` | `uint64_t` | 8 | Total size of the serialized UTXO set in bytes (header + body). |
| `merkle_root` | `uint256` | 32 | Root of the Merkle tree computed over chunk hashes. |

A requesting node MUST ignore entries whose `serialized_hash` does not match a known

> utxo set hash for the corresponding height.

Reviewer 3: I think it would be better for a client that supports this mechanism to hardcode the merkle root instead of the straight serialized hash, and to drop `serialized_hash` from this message.

Author: Hm, yeah, I think that makes sense, I was struggling with finding the right path between building on top of the existing assumeutxo data we already have and extending it. I am adding the merkle root to the assumeutxo data in my PR, so checking the `serialized_hash` is a belts and suspenders there, so it makes sense to make this explicit here as well.

Reviewer 3: If you wanted to change the contents of the utxo dump as well, it would be nice to include the header chain for the block where the snapshot was taken. Then you could do an import without needing to connect to the network at all, I think.

Author: > I think it would be better for a client that supports this mechanism to hardcode the merkle root instead of the straight serialized hash, and to drop serialized_hash from this message.

Dropped `serialized_hash` mentions from the BIP completely in this new version.

> it would be nice to include the header chain for the block where the snapshot was taken

Hm, I find this a bit odd. For the purpose of this BIP we are connected to the network necessarily, so we can just as well fetch the headers over the network in the way we usually do and that should be simpler code-wise. For loading the file directly it would make more sense to me but still we are connecting to the network afterwards. I will need to think about this more but thankfully we have the versioning of the dump file so we can extend the format easily in the future and it wouldn't even need to affect this BIP I think.

#### `getutxoset`

Sent to request a single chunk of UTXO set data. The requesting node MUST have completed header sync
before sending this message.

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `height` | `uint32_t` | 4 | Block height of the requested UTXO set. |
| `block_hash` | `uint256` | 32 | Block hash at the requested height. |
| `chunk_index` | `uint32_t` | 4 | Zero-based index of the requested chunk. |

If the serving node cannot fulfill the request, it MUST NOT respond. The requesting node SHOULD apply
a reasonable timeout and disconnect peers that fail to respond.

#### `utxoset`

Sent in response to `getutxoset`, delivering one chunk with its Merkle proof.

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `height` | `uint32_t` | 4 | Block height this data corresponds to. |
| `block_hash` | `uint256` | 32 | Block hash this data corresponds to. |
| `chunk_index` | `uint32_t` | 4 | Zero-based index of this chunk. |
| `proof_length` | `compact_size` | 1–9 | Number of hashes in the Merkle proof. |

> | `proof_hashes` | `uint256[]` | 32 × `proof_length` | Sibling hashes from leaf to root. |

Reviewer 3: Rather than a proof, it might be better to just request `getutxoset <height> <hash> 0xFFFFFFFF` once to get the full list of chunk hashes -- that should be about 74kB for a 9GB utxo set, and should stay under 4MB until the utxo set is >450GB. Then each chunk is just `<height> <hash> <number> <data>`.

Author: Huh, interesting idea to get all the chunk hashes first, I didn't think of that. It might make sense to do this with a separate message type even instead of hacking `getutxoset` as you described. I will have to think about it a little more.

Author: Specified the fetching of the chunk hash list now with new message type `getutxotree`/`utxotree` messages. I liked this a lot more when combining it with dropping the discovery approach and asking for the specific UTXO set directly. Let me know what you think.

| `data` | `bytes` | variable | Chunk payload, exactly 3.9 MB except for the last chunk. |

The transfer is receiver-driven: the requesting node sends one `getutxoset` per chunk. Chunks MAY be
requested in any order and from different peers, provided those peers advertised the same `merkle_root`
for the same height and block hash.

Upon receiving a `utxoset` message, the node MUST compute `SHA256d(data)` and verify it against the
`merkle_root` using the provided proof. If verification fails, the node MUST discard the chunk and
disconnect the peer. A node SHOULD also disconnect a peer that sends a `utxoset` message with fields
(`chunk_index`, `height`, `block_hash`) that do not match the outstanding request.

After all chunks have been received, the node MUST compute the serialized hash and compare it against a
known UTXO set hash. If this check fails, the node MUST discard all data and

> SHOULD disconnect all peers that advertised the corresponding Merkle root.

Reviewer 5: I'm trying to understand why we "should" disconnect here, but we "must" disconnect if the peer sends us a `utxoset` message that fails the proof verification (L176). Is it because in the second case, the peer is surely trying to trick us, while in the first case, we probably have a bug on our end?

I'm trying to understand how it can happen that our chunks pass the individual verifications, but the final utxo set hash mismatches, and whether in this case it would be a problem on our end, or peers trying to trick us.

Author: Yeah, the intention was that we are stricter with a clear inconsistency in what the peer has given us vs. something that is probably a problem on our side. But have revisited the disconnection policies and I think it is a lot more consistent now. If I got everything right we must disconnect if the peer sent us something clearly inconsistent, we should disconnect if the peer sent us we didn't ask for and if evidence points to the issue being on our side I am now not saying anything about disconnecting.

### Protocol Flow

1. The requesting node identifies peers advertising `NODE_UTXO_SET`.
2. The requesting node sends `getutxosetinfo` to one or more of these peers.
3. Each peer responds with `utxosetinfo`. The requesting node verifies that the advertised
   `serialized_hash` matches a known UTXO set hash, compares `merkle_root` values across peers,

>    and selects a UTXO set whose Merkle root has agreement among multiple peers.

Reviewer 6: I think this process defeats the point of the service bit. Having "at least one" UTXO set only "works" while there are only a small number of UTXO set options to download. If it's not sufficient to just try peers until you find the UTXO set you want, then we probably need a way to advertise *specific* UTXO sets.

Author: Thanks, I do think it should be sufficient to try peers until we find the UTXO set we want and I dropped the discovery approach, see my other comment as well.

4. The requesting node downloads chunks via `getutxoset`/`utxoset` exchanges, verifying each chunk
   against the Merkle root on receipt. On verification failure the peer is disconnected and download
   continues from another peer without losing already-verified chunks.
5. After all chunks are received, the node computes the full serialized hash and verifies it against
   the known UTXO set hash.

Serving nodes are free to limit the number of concurrent and repeated transfers per peer at their own
discretion to manage resource consumption.

## Rationale

**Usage of service bit 12:** Service bits allow selective peer discovery through
DNS seeds and addr relay. Bit 12 is chosen as the next unassigned bit after `NODE_P2P_V2` (bit 11, BIP 324).

**Serialized hash in `utxosetinfo`:** The requesting node should have access to a known UTXO set hash
before initiating the process. Including the serialized hash in the advertisement lets the requester
immediately filter out peers claiming a different UTXO set state before downloading any data.

> **Discovery before download:** The `getutxosetinfo`/`utxosetinfo` exchange lets the requesting node

Reviewer 6: I'm not sure discovery is a useful feature. It enables a misguided implementation to blindly trust whatever nodes provide, and harms node privacy by adding a way to fingerprint nodes.

If you know what you'll accept in advance, you can simply request that snapshot, and the node will either provide it or (potentially) say it can't.

Author: I dropped the discovery approach, I think fingerprinting potential is still there with some probing but it's still better than with discovery and I think this matches the approach we follow in the network generally, like in BIP157 for example, better as well.

transfer.

**Per-chunk Merkle verification:** In the Bitcoin P2P protocol, every larger piece of data received during
normal operation (blocks, transactions, compact block filters) can be verified independently before
requesting more. Without per-chunk verification, a UTXO set transfer would be an anomaly: ~10 GB (as of early 2026)
of data verifiable only after complete receipt. The Merkle tree enables incremental verifiability, allowing for
immediate detection of corrupt data, peer switching without data loss, and parallel download from
multiple peers. The overhead is minimal (~384 bytes of proof per 3.9 MB chunk). The specified
serialization is deterministic, so all honest nodes produce byte-identical output, guaranteeing Merkle
root agreement.

**3.9 MB chunk size:** The number balances round trips (~2,560 for a ~10 GB set) against memory usage
for buffering and verifying a single chunk. Smaller chunks would increase protocol overhead; larger
chunks would increase memory pressure on constrained devices commonly used to run Bitcoin nodes.
Together with the additional message overhead, the `utxoset` message including the chunk data also

> sits just below the theoretical maximum block size which means any implementation should be able to

Reviewer 7: It also happens to sit just below the maximum P2P message size MAX_PROTOCOL_MESSAGE_LENGTH, so it may be clearer to refer to that instead of block size

Author: Yeah, but this was a contious decision actually. `MAX_PROTOCOL_MESSAGE_LENGTH` is a Bitcoin Core implementation specific value. A different implementation may have a higher value for this. But every implementation will at least need to be able to receive the biggest possible block. So I think it's better to anchor it to that.

```

>   BIP: ?
>   Layer: Peer Services
>   Title: P2P UTXO Set Sharing
>   Authors: Fabian Jahr <fjahr@protonmail.com>
>   Status: Draft
>   Type: Specification
>   Assigned: ?

Reviewer: ```suggestion
  BIP: 452
  Layer: Peer Services
  Title: P2P UTXO Set Sharing
  Authors: Fabian Jahr <fjahr@protonmail.com>
  Status: Draft
  Type: Specification
  Assigned: 2026-06-04
```

P2P messages (`getutxotree`, `utxotree`, `getutxoset`, `utxoset`) negotiated via a BIP 434 feature,
and a chunk-hash list anchored to a Merkle root known to the requesting node, enabling per-chunk
verification. This allows bootstrapping nodes to leapfrog to a recent height by obtaining the
required UTXO set directly from the P2P network via mechanisms such as assumeutxo.

## Motivation

The assumeutxo feature (implemented in Bitcoin Core) allows nodes to begin operating from a serialized
UTXO set while validating
historical blocks in the background. However, there is currently no canonical source for obtaining this
data. Users must either generate one themselves from a fully synced node (using `dumptxoutset` in 
Bitcoin Core), or download one from a third party.

By enabling UTXO set sharing over the P2P network, new nodes can obtain the data directly from
peers, removing the dependency on external infrastructure.

## Specification

The key words "MUST", "MUST NOT", "SHOULD", "SHOULD NOT", and "MAY" in this document are to be
interpreted as described in RFC 2119.

### Service Bit

| Name | Bit | Description |
|------|-----|-------------|

> | `NODE_UTXO_SET` | 14 (0x4000) | The node serves complete UTXO set data for the scheduled heights (see [Scheduled UTXO Set Heights](#scheduled-utxo-set-heights)). |

Reviewer 3: Should setting this bit also indicate that you are able to serve blocks between H2 and the tip? That is up to ~30k blocks which might be 60GGB to 120GB vs NODE_NETWORK_LIMITED which only guarantees 288 blocks. Or will nodes just need to find NODE_NETWORK peers to finish IBD in general?

(At height N=961631, H2 is 931392 which is 30,239 blocks earlier)

Reviewer: I don’t think this proposal needs to solve that, when `NODE_NETWORK` already solves it.

sets at a deterministic schedule of block heights derived from the current tip. All heights refer to
blocks on the active most-work chain.

Let:

* `N` = height of the current chain tip
* `M = N - 2016` (the height as of approximately two weeks ago)
* `K = 14112` (7 difficulty adjustment periods, approximately three months)
* `H1 = M - (M mod K)` (the most recent multiple of `K` that is buried by at least 2016 blocks)
* `H2 = H1 - K` (the preceding multiple of `K`)

A node advertising `NODE_UTXO_SET` MUST be able to serve the UTXO sets at heights `H1` and `H2` as
computed from the current tip. Serving two consecutive scheduled heights guarantees an overlap
window: when a new height becomes `H1`, the previous one remains available as `H2`, so an in-progress
download is not interrupted.

The 2016-block offset ensures a height becomes scheduled only once it is buried by approximately two
weeks, which should make it safe from reorganization and gives serving nodes time to produce the
snapshot before it is requested.

### Feature Negotiation

Support for the messages in this document is negotiated per connection via the BIP 434 `feature`

> message, using `featureid` `BIPXXXX` (TODO) and empty `featuredata`. A node implementing these

Reviewer: I thought that the `featuredata` here would include the heights that the serving node can serve. 

If all nodes that implement the protocol send this feature message, the only other way to find out whether a node can serve the release-hardcoded heights would be by asking `getutxotree` or `getutxoset` for a chunk of the corresponding height. What’s the rationale for it being empty? Is it meant to make it more expensive for surveillants to learn what heights each node serves?

If a user wants to download the UTXO set for the trusted hash hardcoded in the latest release, how do they find peers that serve it?

Reviewer 3: I think the idea is that:

 * the node has trusted heights hardcoded with the corresponding block and utxo set hash, eg `935'000`, `0000000000000000000147034958af1652b2b91bba607beacc5e72a56f0fb5ee`, `e4b90ef9eae834f56c4b64d2d50143cee10ad87994c614d7d04125e2a6025050`
 * the node downloads the header the block headers to make sure the block is an ancestor of the tip (and is sufficiently recent that the utxo set is expected to be available)
 * the node uses the service bit to find peers that likely have the utxo set available
 * the node requests the utxoset details by hash, then requests utxoset chunks
 * the node maintains connections to a sufficient number of peers that respond to those chunk requests to keep downloads efficient

So knowing the heights doesn't add anything useful there, I think -- at best it saves you maybe a 100 bytes of message data before you realise they can't send you chunks?

messages advertises this feature and MUST NOT send any of them to a peer that has not.

Advertising the feature indicates only that a node implements the protocol while `NODE_UTXO_SET`
indicates it additionally serves the scheduled heights. A node setting `NODE_UTXO_SET` MUST also
advertise the feature.

### Data Structures

#### Serialized UTXO Set

The serialized UTXO set uses the format established by the Bitcoin Core RPC `dumptxoutset` (as of Bitcoin Core v31).

**Header (55 bytes):**

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `magic` | `bytes` | 5 | `0x7574786fff` (ASCII `utxo` + `0xff`). |

> | `version` | `uint16_t` | 2 | Format version. |

Reviewer 3: Should specify the value (`0x0200`?) here? This is effectively standardizing/documenting bitcoin core's current format.

order of the coins database.

#### Chunk Merkle Tree

The serialized UTXO set (header + body) is split into chunks of exactly 3,900,000 bytes (3.9 MB). The
last chunk contains the remaining bytes and may be smaller. The chunks form the leaves of a binary
Merkle tree whose root commits to the entire UTXO set.

The leaf hash for each chunk is `SHA256d(chunk_data)`. The tree is built as a balanced binary tree. When
the number of nodes at a level is odd, the last node is promoted unchanged to the next level.
Interior nodes are computed as `SHA256d(left_child || right_child)`.

The leaves are delivered to the node in a single `utxotree` response. A node that knows
the Merkle root for a given UTXO set checks a received list of leaves by recomputing the root and
comparing. The Merkle root is the sole trust input required to verify the integrity of the received UTXO set.

`SHA256d` denotes double-SHA256: `SHA256d(x) = SHA256(SHA256(x))`.

### Messages

#### `getutxotree`

Sent to request the chunk-hash list for a specific UTXO set.

| Field | Type | Size | Description |
|-------|------|------|-------------|

> | `block_hash` | `uint256` | 32 | Block hash identifying the requested UTXO set. |

Reviewer 3: I think `getutxotree` should be taking a "chunk merkle tree root hash" as input, rather than a `block_hash`. You want the utxotree that matches the hash that was hardcoded into your node software, not any random utxo serialization that a peer might have. Particularly applies if the utxotree format changes in future -- even an honest peer with the same utxo set won't be helpful if they give you the utxos in a different format that you can't decode correctly.

A node that has advertised `NODE_UTXO_SET` and can serve the requested UTXO set MUST respond with
`utxotree`. If the serving node cannot fulfill the request, it MUST NOT respond. The requesting

> node SHOULD apply a reasonable timeout and try another peer.

Reviewer 3: I think a notfound message would be better here, really. Failing that, perhaps the serving node should disconnect, rather than simply not responding? If you're trying to obtain your first utxo set, and this connection can't do it, there's not much benefit to either party to keeping the connection open is there? Likewise for inability to respond to `getutxoset` requests.

#### `utxotree`

Sent in response to `getutxotree`, delivering the full chunk-hash list along with per-snapshot
metadata.

| Field | Type | Size | Description |
|-------|------|------|-------------|
| `block_hash` | `uint256` | 32 | Block hash this data corresponds to. |

> | `version` | `uint16_t` | 2 | Format version of the serialized UTXO set. |

Reviewer 3: These are redundant -- you're already checking them when you verify the chunk_hahses match the hardcoded hash.

| `data_length` | `uint64_t` | 8 | Total size of the serialized UTXO set in bytes (header + body). |
| `chunk_hashes` | `uint256[]` | 32 × N | The ordered list of N chunk hashes, where N = `ceil(data_length / 3,900,000)`. |

Upon receiving a `utxotree` message, the requesting node MUST recompute the Merkle root from
`chunk_hashes` and compare it against the Merkle root it knows for the corresponding UTXO set. If

> the roots do not match, the node MUST discard the response and MUST disconnect the peer.

Reviewer 3: I believe that if you don't include `data_length` in the merkle root, then you get ambiguity between (eg) 1000 chunks of 3.9GB of total data and 999 chunks of 3,892,200,064 bytes of data, where the final 64 byte chunk is the concatenated hashes of the original final two chunks.

#### `getutxoset`

Sent to request a single chunk of UTXO set data. The requesting node MUST have received a `utxotree`
for the corresponding UTXO set (from any peer) before sending this message.

| Field | Type | Size | Description |
|-------|------|------|-------------|

> | `block_hash` | `uint256` | 32 | Block hash identifying the requested UTXO set. |

Reviewer 3: Should be the chunk merkle root hash -- what if you're serving utxo sets for block B in both format 2 and some future format 3?
